File: hashingwithcollision.py
# class HashTable:
#     def __init__(self, size):
#         self.size = size
#         self.table = [None] * size

#     def insert(self, key, value):
#         index = hash(key) % self.size
#         if self.table[index] is None:
#             self.table[index] = (key, value)
#         else:
#             # Collision! Try the next index
#             print("Collision:", key, value, "at index", index)
#             next_index = (index + 1) % self.size
#             while self.table[next_index] is not None:
#                 print("Collision:", key, value, "at index", next_index)
#                 next_index = (next_index + 1) % self.size
#             self.table[next_index] = (key, value)

#     def lookup(self, key):
#         index = hash(key) % self.size
#         if self.table[index] is None:
#             return None
#         elif self.table[index][0] == key:
#             return self.table[index][1]
#         else:
#             # Check for collision
#             next_index = (index + 1) % self.size
#             while self.table[next_index] is not None and self.table[next_index][0] != key:
#                 next_index = (next_index + 1) % self.size
#             if self.table[next_index] is None:
#                 return None
#             else:
#                 return self.table[next_index][1]
 
def print_collisions(hash_table): 
    for i, basket in enumerate(hash_table): # Loop over each basket in the hash table
        if len(basket) > 1: # If there's more than one item in the basket
            print(f"Collision index {i}: {basket}") # Print the basket

hash_table = [[] for _ in range(10)] # Create a list of 10 empty lists
 
# Add some values to the hash table
hash_table[3].append("grape") # Add "grape" to the basket at index 3
hash_table[4].append("apple") # Add "apple" to the basket at index 5
hash_table[4].append("banana") # Add "banana" to the basket at index 5
hash_table[5].append("orange") # Add "orange" to the basket at index 7
hash_table[5].append("kiwi") # Add "kiwi" to the basket at index 7

print_collisions(hash_table)

File: hashingwithcollision.java
import java.util.ArrayList;
import java.util.List;

public class hashingwithcollision { 

    public static void printCollisions(List<List<String>> hashTable) { 
        for (int i = 0; i < hashTable.size(); i++) { // iterate through the hash table
            List<String> basket = hashTable.get(i); // get the basket at index i 
            if (basket.size() > 1) { // if the basket has more than one item, there is a collision
                System.out.println("Collision index " + i + ": " + basket); 
            }
        }
    }

    public static void main(String[] args) { 
        List<List<String>> hashTable = new ArrayList<List<String>>(); // create a hash table
        for (int i = 0; i < 10; i++) { // add 10 baskets to the hash table
            hashTable.add(new ArrayList<String>()); // each basket is an empty list
        } 

        hashTable.get(3).add("grape");
        hashTable.get(4).add("apple");
        hashTable.get(4).add("banana");
        hashTable.get(5).add("orange");
        hashTable.get(5).add("kiwi");

        printCollisions(hashTable);
    }
}
